### ISSUE TITLE
Change national reweighting objective function and replace PyTorch L-BFGS with Clarabel QP solver

### ISSUE THREAD

**Issue by donboyd5** [PSLmodels/tax-microdata-benchmarking#415]
## Problem

The current reweighting uses a penalty-based objective that combines target accuracy and weight smoothness into a single sum. It imposes bounds on how much individual weights can change from their pre-optimization values. It solves this using PyTorch L-BFGS. The objective function, and how it is solved, has two structural problems:

1. **No per-target accuracy guarantee.** The optimizer minimizes the total penalty, so hard-to-satisfy targets can drift far from their SOI values if that reduces total loss. Unemployment compensation targets, for example, are ~8% off in the current weights.

2. **Cross-machine non-reproducibility.** Issue #400 and prior work (PRs #407, #411) improved this substantially, but PyTorch L-BFGS with `torch.clamp()` bounds still has inherent non-determinism across different hardware (flat gradient regions from clamping, BLAS differences, float32 reduction order).

Issue #413 proposed switching to scipy L-BFGS-B, which reduced the reproducibility problem by using projected-gradient bounds and analytical gradients, but cross-machine determinism still depended on having the same BLAS library, and the flat objective surface near the optimum remained a concern. It also kept the same penalty-based objective, so the per-target accuracy issue remained.

## Proposal

Replace the penalty-based formulation with an objective that minimizes changes in weights subject to satisfying SOI target constraints. This constrained formulation is common in microdata reweighting — the [PSL taxdata project](https://github.com/PSLmodels/taxdata) and others use variants where targets are constraints rather than penalties. The current tmd approach, like PolicyEngine's, uses a target-minimization penalty objective.

Structure this as a **constrained quadratic program** solved by the [Clarabel](https://oxfordcontrol.github.io/ClarabelDocs/) interior-point solver:

```
minimize    sum((x_i - 1)^2)                              [weight distortion]
subject to  target_j * (1 - eps) <= achieved_j <= target_j * (1 + eps)   [+/-0.5%]
            x_min <= x_i <= x_max                          [multiplier bounds]
```

Key differences from the current approach:

- **Targets are hard constraints, not penalties.** Every SOI target must be satisfied within +/-0.5% (configurable). No target can be sacrificed to improve others.
- **Deterministic across machines.** Clarabel is a Rust-based interior-point solver operating in float64. Interior-point methods converge to a unique optimum determined by the KKT conditions, not by reduction order.
- **Elastic slack variables** for graceful infeasibility handling (see below).
- **Faster.** Solve time on an AMD Ryzen 9 9950X machine is ~10 seconds (vs ~7 min scipy L-BFGS-B, ~90 seconds PyTorch L-BFGS with NVIDIA GeForce RTX 5070 Ti GPU).
- **No GPU or PyTorch dependency** for the reweighting step.

### Elastic slack variables

A constrained optimization problem can fail if any constraint is infeasible — that is, if no combination of weight multipliers can bring a target within the required tolerance band. Without slack variables, the solver would simply report "infeasible" and return no solution at all.

Elastic slack variables prevent this. For each target constraint, two non-negative slack variables (s_lo, s_hi) are added that can widen the constraint bounds:

```
lb_j <= achieved_j + s_lo_j - s_hi_j <= ub_j       s_lo, s_hi >= 0
```

These slacks are penalized heavily in the objective (penalty M = 1e6 by default):

```
minimize    sum((x_i - 1)^2) + M * sum(s^2)
```

The large penalty means the solver tries very hard to satisfy all constraints without slack (driving s toward zero). But if a constraint is geometrically impossible — for example, if all data values for a target are zero, so no reweighting can produce a nonzero estimate — the solver uses just enough slack to make the problem feasible rather than failing entirely.

After solving, the diagnostics report which constraints (if any) required slack and how much, making it easy to identify problematic targets. In practice, no slack was needed — all ~550 constraints were satisfied within the +/-0.5% tolerance band without any relaxation.

### Why Clarabel

Clarabel was selected after testing six QP/NLP solvers:

| Solver | Type | Result |
|--------|------|--------|
| **OSQP** | First-order (ADMM) | Converges quickly to rough solutions but cannot reach tight tolerances — a fundamental limitation of first-order methods |
| **PIQP** | First-order | Diverges, cannot handle the problem structure |
| **ProxQP** | Proximal interior-point | Hung without making progress |
| **HiGHS** | Active-set | Very slow convergence; warned about large coefficient ranges |
| **IPOPT** | Interior-point (NLP) | Converged to high accuracy but slow (~820s) and requires system library (MUMPS) |
| **Clarabel** | Interior-point (QP) | Converged in ~10 seconds to high accuracy; all constraints satisfied; deterministic |

Interior-point methods are well-suited to this problem: moderate size (~225K variables, ~550 constraints), need for high accuracy, and a sparse constraint matrix. Clarabel achieves the same solution quality as IPOPT but is ~57x faster, and only requires `pip install clarabel` (no system libraries).

## Results

All ~550 SOI targets are satisfied within the 0.5% tolerance band. Previously problematic targets (unemployment compensation at ~8% error) are now within 0.5%. Cross-machine reproducibility has been confirmed: identical weight fingerprints on two different machines (different CPUs, different OS configurations).

## Backward compatibility

The PyTorch and scipy solvers are preserved and selectable via environment variables (`PYTORCH_REWEIGHT=1`, `SCIPY_REWEIGHT=1`). Clarabel becomes the default.
